db/migrate/20240903123622_back_fill_mentor_training_training_type.rb

```diff
@@ -0,0 +1,8 @@
+class BackFillMentorTrainingTrainingType < ActiveRecord::Migration[7.1]
+  def up
+    # We require callbacks for BigQuery to stay in sync with the data.
+    Claims::MentorTraining.find_each do |mentor_training|
+      mentor_training.update(training_type: Claims::MentorTraining.training_types[:initial])
+    end
+  end
+end
```

Comment:
I believe we've decided to keep data migrations outside of schema migrations. That being said, you'll need to add the data migration in a separate PR to ensure that the `null` db values are cleared out before the validations are added.

Reply:
I've had a conversation with @ollietreend about the removal of data migration. I'm all on board with the approach.

One thing important to me is the re-playability of migrations though. In the case here, we use application code within the migration file. As application code is prone to change, for example:

2 years from now, we decide to rename the `Claims::MentorTraining` model. This breaks the migration. Alternatively, we could change the `training_type` enum name or values. This could also break the migration.

The workaround is to avoid using application code from within migrations. Executing raw SQL is a sure way to ensure that no application code is used, but as a Rails developer, `ActiveRecord` is available for a reason, no? 😉 

```ruby
class SomeMigration < ActiveRecord::Migration[7.1]
  class MentorTrainingForMigration < ActiveRecord::Base # Note that this doesn't inherit from ApplicationRecord
    self.table_name = :mentor_trainings
  end

  def up
    MentorTrainingForMigration.update_all(training_type: :initial)
  end

  def down
    MentorTrainingForMigration.update_all(training_type: nil)
  end
end
```

As for pushing the updated values from our Prod database to DfE Analytics, let's perform a manual update following the guidance [here](https://github.com/DFE-Digital/dfe-analytics?tab=readme-ov-file#7-import-existing-data).

Better yet, you can tie this data migration to the schema change too.

```ruby
class ChangeMentorTrainingTrainingTypeDefault < ActiveRecord::Migration[7.1]
  class MentorTrainingForMigration < ActiveRecord::Base
    self.table_name = :mentor_trainings
  end

  def change
    change_column_default :mentor_trainings, :training_type, from: nil, to: "initial"

    up_only do
      MentorTrainingForMigration.where(training_type: nil).update_all(training_type: :initial)
    end
  end
end
```

Note: This would update the values when migrating forward, but leaves them as is when reverting.

https://github.com/testdouble/good-migrations is probably something we can add to prevent application code from slipping into migrations.

Reply:
Some great insights @Nitemaeric thank you, I've done similar previously when a model hasn't been available so this is a great idea 👍